sticker/src/model/mod.rs

```diff
@@ -0,0 +1,243 @@
+    ///
+    /// `layer_dropout` is the probability with which layers should
+    /// be dropped out in scalar weighting during training.
+    pub fn new<'a>(
+        vs: impl Borrow<Path<'a>>,
+        config: &BertConfig,
+        encoders: &Encoders,
+        layers_dropout: f64,
+    ) -> Result<Self, BertError> {
+        let vs = vs.borrow();
+
+        let embeddings = BertEmbeddings::new(vs.sub("encoder"), config);
+        let encoder = BertEncoder::new(vs.sub("encoder"), config)?;
+
+        let classifiers = encoders
+            .iter()
+            .map(|encoder| {
+                (
+                    encoder.name().to_owned(),
+                    ScalarWeightClassifier::new(
+                        vs.sub("classifiers")
+                            .sub(format!("{}_classifier", encoder.name())),
+                        config.num_hidden_layers,
+                        config.hidden_size,
+                        encoder.encoder().len() as i64,
+                        0.1,
+                    ),
+                )
+            })
+            .collect();
+
+        Ok(BertModel {
+            classifiers,
+            embeddings,
+            encoder,
+            layers_dropout: Dropout::new(layers_dropout),
+        })
+    }
+
+    /// Construct a model and load parameters from a pretrained model.
+    ///
+    /// `layer_dropout` is the probability with which layers should
+    /// be dropped out in scalar weighting during training.
+    pub fn from_pretrained<'a>(
+        vs: impl Borrow<Path<'a>>,
+        config: &BertConfig,
+        hdf_path: impl AsRef<path::Path>,
+        encoders: &Encoders,
+        layers_dropout: f64,
+    ) -> Fallible<Self> {
+        let vs = vs.borrow();
+
+        let pretrained_file = File::open(hdf_path, "r")?;
+
+        let embeddings = BertEmbeddings::load_from_hdf5(
+            vs.sub("encoder"),
+            config,
+            pretrained_file.group("bert/embeddings")?,
+        )?;
+
+        let encoder = BertEncoder::load_from_hdf5(
+            vs.sub("encoder"),
+            config,
+            pretrained_file.group("bert/encoder")?,
+        )?;
+
+        let classifiers = encoders
+            .iter()
+            .map(|encoder| {
+                (
+                    encoder.name().to_owned(),
+                    ScalarWeightClassifier::new(
+                        vs.sub("classifiers")
+                            .sub(format!("{}_classifier", encoder.name())),
+                        config.num_hidden_layers,
+                        config.hidden_size,
+                        encoder.encoder().len() as i64,
+                        0.1,
+                    ),
+                )
+            })
+            .collect();
+
+        Ok(BertModel {
+            embeddings,
+            encoder,
+            layers_dropout: Dropout::new(layers_dropout),
+            classifiers,
+        })
+    }
+
+    /// Encode an input.
+    fn encode(
```

Comment:
Wouldn't `forward_t` be the fitting name for this?

Reply:
But encode only produces hidden states. I would expect a `forward_t` method to also apply the scalar weighting and perhaps softmax.

Reply:
Ah I see. Still a bit undecided though: from a `BertModel::forward_t` I'd expect the bert encoder and not scalar weighted classification.